File: R/recursiveBS.r
# Function for binary search
recursivebinarySearch <- function(vector, low, high, target) {
  if(low > high){
      return(-1)
  }

  mid <- floor((low + high) / 2)
  if(vector[mid] == target){
      return(mid)
  } else if (vector[mid] < target) {
      return(recursivebinarySearch(vector, mid + 1, high, target))
  } else {
      return(recursivebinarySearch(vector, low, mid - 1, target))
  }
}

# Read command line arguments
args <- commandArgs(trailingOnly = TRUE)

# Validate and parse arguments
if (length(args) >= 2) {
  vector <- as.integer(unlist(strsplit(args[1], ",")))
  target <- as.integer(args[2])

  # Perform binary search
  index <- recursivebinarySearch(vector, 1, length(vector), target)

  # Output the result
  if (index != -1) {
    cat("Target found at index:", index, "\n")
  } else {
    cat("Target not found in the vector.\n")
  }
} else {
  cat("Usage: Rscript binary_search.R \"1,3,5,7,9\" 5\n")
}

File: cpp/recursiveBS.cpp
#include <iostream>
#include <algorithm>
#include<vector>


int recursiveBS(const std::vector<int>& arr,int low,int high,int target){
 
 if(low == high){
    if(arr[low] ==target){
        return low;
    }
    else{
        return -1;
    }
 }
 else{
    int mid= (low+(high-low))/2;
    if(arr[mid] == target){
        return mid;
    }
    else if (arr[mid]>target){
        return recursiveBS(arr,mid+1,high,target);
    }
    else{
        return recursiveBS(arr,low,mid-1,target);
    }
 }

}


int main() {
    int n, target;
    std::cout << "Enter number of elements: ";
    std::cin >> n;

    std::vector<int> arr(n);
    std::cout << "Enter " << n << " elements in sorted order:\n";
    for (int i = 0; i < n; ++i) {
        std::cin >> arr[i];
    }

    std::cout << "Enter the element to search for: ";
    std::cin >> target;

    // Optionally sort the array if the input is not guaranteed to be sorted
    // std::sort(arr.begin(), arr.end());

    int result = recursiveBS(arr,0,n,target);

    if (result != -1) {
        std::cout << "Element found at index: " << result << std::endl;
    } else {
        std::cout << "Element not found in the array." << std::endl;
    }

    return 0;
}
